### PR TITLE
Save the letter to Sent folder by default #1

### PR DESCRIPTION
**Pull request by hanxianlong**
I think it would be better to save letter to sender folder by default,
So I added the the following mailet to mailetcontainer.xml
`<mailet match="SenderIsLocal" class="ToSenderFolder">
                <folder>Sent</folder>
                <consume>false</consume>
            </mailet>`
It did save the letter to Sent folder, **But there is another issue: when I sent the letter to local domain, and cc to other domain, it saved two letters to the Sent folder, which were the same.**
So I made the change to ToSenderFolder mailet, ignored the mail's state when service method is called, and then put the ToSenderFolder mailet behind the HostIsLocal mailet.

`@Override
    public void service(Mail mail) throws MessagingException {
           doService(mail);
            if (consume) {
                mail.setState(Mail.GHOST);
            }
    } `

`<processor state="transport" enableJmx="true">
...
            <mailet match="HostIsLocal" class="ToProcessor">
                <processor>local-address-error</processor>
                <notice>550 - Requested action not taken: no such user here</notice>
            </mailet>

            <mailet match="SenderIsLocal" class="ToSenderFolder">
                <folder>Sent</folder>
                <consume>false</consume>
            </mailet>

            <mailet match="relay-allowed" class="ToProcessor">
                <processor>relay</processor>
            </mailet>
...
        </processor>`

It works well for me, hopes it can help the other people.

### CODE REVIEW

File: dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml [hunk 1]
```diff
@@ -78,6 +78,11 @@
                 <notice>550 - Requested action not taken: no such user here</notice>
             </mailet>
 
+	    <mailet match="SenderIsLocal" class="ToSenderFolder">
+	    	<folder>Sent</folder>
+		<consume>false</consume>
+	    </mailet>
```
Comment (chibenwa) [inline dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml:84]: Generally the mua takes care of saving stuff in sent folder.

While I agree about improving the doc, I am unsure this change should make it to default configuration...

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConnectionFactory.java [hunk 2]
```diff
@@ -50,6 +50,10 @@ private ConnectionFactory from(RabbitMQConfiguration rabbitMQConfiguration) {
             connectionFactory.setChannelRpcTimeout(rabbitMQConfiguration.getChannelRpcTimeoutInMs());
             connectionFactory.setConnectionTimeout(rabbitMQConfiguration.getConnectionTimeoutInMs());
             connectionFactory.setNetworkRecoveryInterval(rabbitMQConfiguration.getNetworkRecoveryIntervalInMs());
+
+            connectionFactory.setUsername(rabbitMQConfiguration.getManagementCredentials().getUser());
+            connectionFactory.setPassword(String.valueOf(rabbitMQConfiguration.getManagementCredentials().getPassword()));
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConnectionFactory.java:55]: Unrelated change?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java [hunk 3]
```diff
@@ -71,12 +71,10 @@ public ToSenderFolder(UsersRepository usersRepository, @Named("mailboxmanager")
      */
     @Override
     public void service(Mail mail) throws MessagingException {
-        if (!mail.getState().equals(Mail.GHOST)) {
```
[the remark this answers is not in the document]
Comment (hanxianlong) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java:74]: With cassandra distributed version: I'd like to save the sent letter to Sent folder. When I add the ToSenderFolder to the mailetcontainer(as shown in the commit), it doesn't work correctlly, the letter couldn't be saved in the Sent folder because of the mail state becomes to Ghost.
I think if someone want's to save the letter to Sent,  we shouldn't care the mail's sate, no matter what sate it is, only if  the ToSendFolder.doService methods is called, it should work. I'm not sure what I think is correct, what's your opion? @chibenwa

[the remark this answers is not in the document]
Comment (hanxianlong) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java:74]: Or do you have any other better solution to solve the issue?
1. save the letter to sent
2. If  "to" is local domain, "cc" is the other domain, it saves two copies of the letter to Sent folder(which is unacceptable）

File: dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml (hunk 1 above)
Comment (hanxianlong) [reply dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml:84]: [Generally the mua takes care of saving stuff in sent folder.](url)
Think about the gmail/netease, the mail server will save the sent letter to the mail server by default. In my opinion, james should work save as well, If they want to turn it off, they can remove the ToSentFolder, not the other way.

**Review by hanxianlong — COMMENTED** [review]
https://github.com/linagora/james-project/pull/4036#discussion_r523541574
Sorry, This is the previouse commit. I'd like to make another new PR if we can reach an agreement with the issue of _Save the letter to Sent folder_

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java:74]: Have you been configuring LocalDelivery with `<consume>false</consume>` ?

File: dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml (hunk 1 above)
Comment (chibenwa) [reply dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml:84]: I have thunderbird configured on top of gmail, when I send a mail via smtp, there is a second imap request to save the mail in sent folder.

While better documenting how "to sender foler wotks" i really doubt enabling it by default would be a good move.

Cheers,

Benoit

Comment (rouazana) [reply dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/mailetcontainer.xml:84]: > Think about the gmail/netease, the mail server will save the sent letter to the mail server by default. In my opinion, james should work save as well, If they want to turn it off, they can remove the ToSentFolder, not the other way.

I some users want to configure this, it's more convenient to let the configuration of this at the MUA side than on the server side.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java:74]: Also there should be only one email transiting, with "both to and cc positioned", and the envoleope should cary out both To and Cc recipients, ensuring the email is saved only once.

I don't see how your changeset would be chaging anything.

Can you write a test?

File: backends-common/cassandra/apache-james-backends-cassandra.iml [hunk 4]
```diff
@@ -0,0 +1,118 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline backends-common/cassandra/apache-james-backends-cassandra.iml:1]: Don t commit .iml files...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderFolder.java:74]: **No answer here**

**Comment by chibenwa** [comment]
Sorry I still did not get the usefullness of that work...
